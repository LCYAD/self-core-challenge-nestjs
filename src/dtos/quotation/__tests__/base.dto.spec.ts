import { COMMON_FIELD } from '@constants/fakers/common.faker.constant'
import {
  generateInvalidArrayLargerThanMaxSizeTest,
  generateInvalidArraySmallerThanMinSizeTest,
  generateInvalidArrayTest,
  generateInvalidNestedObjectArrayTest,
  generateInvalidNestedObjectTest,
  generateInvalidNumberTest,
  generateInvalidStringFormatTest,
  generateMissingFieldTest,
  generateValidBaseTest,
  generateValidOptionalTest,
  testInvalid
} from '@tests/helper/dto.helper'

import {
  QuotationBaseDto,
  QuotationBaseItemDto,
  quotationBaseItemValidExample,
  quotationBaseValidExample
} from '../base.dto'
import { quotationStopValidExample } from '../stop.dto'

describe('QuotationBaseItemDto', () => {
  describe('valid response', () => {
    generateValidBaseTest(QuotationBaseItemDto, quotationBaseItemValidExample)
    generateValidOptionalTest(
      ['categories', 'handlingInstructions'],
      QuotationBaseItemDto,
      quotationBaseItemValidExample
    )
  })

  describe('invalid response', () => {
    generateMissingFieldTest(
      ['quantity', 'weight'],
      QuotationBaseItemDto,
      quotationBaseItemValidExample
    )
    generateInvalidArrayTest(
      ['categories', 'handlingInstructions'],
      QuotationBaseItemDto,
      quotationBaseItemValidExample
    )
    generateInvalidNumberTest(
      ['quantity', 'weight'],
      QuotationBaseItemDto,
      quotationBaseItemValidExample
    )
  })
})

describe('QuotationBaseDto', () => {
  beforeAll(() => {
    jest.spyOn(Date, 'now').mockReturnValue(COMMON_FIELD.CURRENT_TIME_MILLI_SEC)
  })
  describe('valid response', () => {
    generateValidBaseTest(QuotationBaseDto, quotationBaseValidExample)
  })

  describe('invalid response', () => {
    generateMissingFieldTest(
      [
        'quotationId',
        'expireAt',
        'scheduleAt',
        'deliveryBy',
        'location',
        'stops',
        'item'
      ],
      QuotationBaseDto,
      quotationBaseValidExample
    )
    generateInvalidArrayTest(
      ['stops'],
      QuotationBaseDto,
      quotationBaseValidExample
    )
    generateInvalidStringFormatTest(
      [
        { key: 'quotationId', invalidString: 'xxx' },
        { key: 'expireAt', invalidString: 'xxx' },
        { key: 'scheduleAt', invalidString: 'xxx' },
        { key: 'deliveryBy', invalidString: 'xxx' }
      ],
      QuotationBaseDto,
      quotationBaseValidExample
    )
    generateInvalidArrayLargerThanMaxSizeTest(
      [
        {
          key: 'stops',
          allowedMaxSize: 2,
          data: quotationStopValidExample
        }
      ],
      QuotationBaseDto,
      quotationBaseValidExample
    )
    generateInvalidArraySmallerThanMinSizeTest(
      [
        {
          key: 'stops',
          allowedMinSize: 2,
          data: quotationStopValidExample
        }
      ],
      QuotationBaseDto,
      quotationBaseValidExample
    )
    generateInvalidNestedObjectArrayTest(
      ['stops'],
      QuotationBaseDto,
      quotationBaseValidExample
    )
    generateInvalidNestedObjectTest(
      ['item'],
      QuotationBaseDto,
      quotationBaseValidExample
    )
    it('for deliveryBy earlier than scheduledAt', (done) => {
      testInvalid(done)(QuotationBaseDto, {
        ...quotationBaseValidExample,
        deliveryBy: COMMON_FIELD.CURRENT_TIME,
        scheduleAt: COMMON_FIELD.AFTER_1HR_CURRENT_TIME
      })
    })
  })
})
