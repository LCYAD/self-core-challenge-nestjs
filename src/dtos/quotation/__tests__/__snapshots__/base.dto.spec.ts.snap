// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuotationBaseDto invalid response for deliveryBy earlier than scheduledAt 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "IsLaterThan": "deliveryBy must be later than scheduleAt, now",
    },
    "property": "deliveryBy",
    "value": "2022-05-21T00:00:00.000Z",
  },
]
`;

exports[`QuotationBaseDto invalid response for deliveryBy not a allowed value xxx 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "IsLaterThan": "deliveryBy must be later than scheduleAt, now",
      "isDateString": "deliveryBy must be a valid ISO 8601 date string",
    },
    "property": "deliveryBy",
    "value": "xxx",
  },
]
`;

exports[`QuotationBaseDto invalid response for expireAt not a allowed value xxx 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isDateString": "expireAt must be a valid ISO 8601 date string",
    },
    "property": "expireAt",
    "value": "xxx",
  },
]
`;

exports[`QuotationBaseDto invalid response for invalid nested object inside array stops 1`] = `
[
  ValidationError {
    "children": [
      ValidationError {
        "children": [
          ValidationError {
            "children": [],
            "constraints": {
              "isString": "address must be a string",
            },
            "property": "address",
            "value": undefined,
          },
        ],
        "property": "0",
        "value": QuotationStopDto {
          "address": undefined,
          "coordinates": undefined,
        },
      },
    ],
    "constraints": {
      "arrayMinSize": "stops must contain at least 2 elements",
    },
    "property": "stops",
    "value": [
      QuotationStopDto {
        "address": undefined,
        "coordinates": undefined,
      },
    ],
  },
]
`;

exports[`QuotationBaseDto invalid response for item as an invalid nested object 1`] = `
[
  ValidationError {
    "children": [
      ValidationError {
        "constraints": {
          "nestedValidation": "nested property item must be either object or array",
        },
        "property": "item",
        "target": [Function],
        "value": "xxx",
      },
    ],
    "property": "item",
    "value": "xxx",
  },
]
`;

exports[`QuotationBaseDto invalid response for missing deliveryBy 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "IsLaterThan": "deliveryBy must be later than scheduleAt, now",
      "isDateString": "deliveryBy must be a valid ISO 8601 date string",
    },
    "property": "deliveryBy",
    "value": undefined,
  },
]
`;

exports[`QuotationBaseDto invalid response for missing expireAt 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isDateString": "expireAt must be a valid ISO 8601 date string",
    },
    "property": "expireAt",
    "value": undefined,
  },
]
`;

exports[`QuotationBaseDto invalid response for missing item 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isDefined": "item should not be null or undefined",
    },
    "property": "item",
    "value": undefined,
  },
]
`;

exports[`QuotationBaseDto invalid response for missing location 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "location must be a string",
    },
    "property": "location",
    "value": undefined,
  },
]
`;

exports[`QuotationBaseDto invalid response for missing quotationId 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "quotationId must be a string",
      "matches": "quotationId must match /^[0-9]{12}$/ regular expression",
    },
    "property": "quotationId",
    "value": undefined,
  },
]
`;

exports[`QuotationBaseDto invalid response for missing scheduleAt 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "IsLaterThan": "scheduleAt must be later than now",
      "isDateString": "scheduleAt must be a valid ISO 8601 date string",
    },
    "property": "scheduleAt",
    "value": undefined,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "IsLaterThan": "deliveryBy must be later than scheduleAt, now",
    },
    "property": "deliveryBy",
    "value": "2022-05-21T01:00:00.000Z",
  },
]
`;

exports[`QuotationBaseDto invalid response for missing stops 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "arrayMaxSize": "stops must contain not more than 2 elements",
      "arrayMinSize": "stops must contain at least 2 elements",
      "isArray": "stops must be an array",
      "isDefined": "stops should not be null or undefined",
    },
    "property": "stops",
    "value": undefined,
  },
]
`;

exports[`QuotationBaseDto invalid response for quotationId not a allowed value xxx 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "matches": "quotationId must match /^[0-9]{12}$/ regular expression",
    },
    "property": "quotationId",
    "value": "xxx",
  },
]
`;

exports[`QuotationBaseDto invalid response for scheduleAt not a allowed value xxx 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "IsLaterThan": "scheduleAt must be later than now",
      "isDateString": "scheduleAt must be a valid ISO 8601 date string",
    },
    "property": "scheduleAt",
    "value": "xxx",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "IsLaterThan": "deliveryBy must be later than scheduleAt, now",
    },
    "property": "deliveryBy",
    "value": "2022-05-21T01:00:00.000Z",
  },
]
`;

exports[`QuotationBaseDto invalid response for stops has less than 2 of items in array 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "arrayMinSize": "stops must contain at least 2 elements",
    },
    "property": "stops",
    "value": [
      QuotationStopDto {
        "address": "999 Prince Edward Road",
        "coordinates": QuotationStopCoordinatesDto {
          "lat": "22.335313",
          "lng": "114.175840",
        },
      },
    ],
  },
]
`;

exports[`QuotationBaseDto invalid response for stops has more than 2 of items in array 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "arrayMaxSize": "stops must contain not more than 2 elements",
    },
    "property": "stops",
    "value": [
      QuotationStopDto {
        "address": "999 Prince Edward Road",
        "coordinates": QuotationStopCoordinatesDto {
          "lat": "22.335313",
          "lng": "114.175840",
        },
      },
      QuotationStopDto {
        "address": "999 Prince Edward Road",
        "coordinates": QuotationStopCoordinatesDto {
          "lat": "22.335313",
          "lng": "114.175840",
        },
      },
      QuotationStopDto {
        "address": "999 Prince Edward Road",
        "coordinates": QuotationStopCoordinatesDto {
          "lat": "22.335313",
          "lng": "114.175840",
        },
      },
    ],
  },
]
`;

exports[`QuotationBaseDto invalid response for stops not an array 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "arrayMaxSize": "stops must contain not more than 2 elements",
      "arrayMinSize": "stops must contain at least 2 elements",
      "isArray": "stops must be an array",
    },
    "property": "stops",
    "value": QuotationStopDto {
      "address": "999 Prince Edward Road",
      "coordinates": QuotationStopCoordinatesDto {
        "lat": "22.335313",
        "lng": "114.175840",
      },
    },
  },
]
`;

exports[`QuotationBaseItemDto invalid response for categories not an array 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isArray": "categories must be an array",
    },
    "property": "categories",
    "value": "test",
  },
]
`;

exports[`QuotationBaseItemDto invalid response for handlingInstructions not an array 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isArray": "handlingInstructions must be an array",
    },
    "property": "handlingInstructions",
    "value": "test",
  },
]
`;

exports[`QuotationBaseItemDto invalid response for missing quantity 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isNumber": "quantity must be a number conforming to the specified constraints",
    },
    "property": "quantity",
    "value": undefined,
  },
]
`;

exports[`QuotationBaseItemDto invalid response for missing weight 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isNumber": "weight must be a number conforming to the specified constraints",
    },
    "property": "weight",
    "value": undefined,
  },
]
`;

exports[`QuotationBaseItemDto invalid response for quantity not an number 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isNumber": "quantity must be a number conforming to the specified constraints",
    },
    "property": "quantity",
    "value": "xxx",
  },
]
`;

exports[`QuotationBaseItemDto invalid response for weight not an number 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isNumber": "weight must be a number conforming to the specified constraints",
    },
    "property": "weight",
    "value": "xxx",
  },
]
`;
